tests/Application/Actions/Donations/UpdateTest.php

```diff
@@ -1765,4 +1754,242 @@ public function testAddDataSuccessWithCashBalanceAutoconfirm(): void
+        $stripePaymentIntentsProphecy->update(Argument::cetera())
+            ->shouldNotBeCalled();
+        $stripePaymentIntentsProphecy->confirm('pi_externalId_123')
+            ->shouldNotBeCalled();
+        $stripeClientProphecy = $this->prophesize(StripeClient::class);
+        $stripeClientProphecy->paymentIntents = $stripePaymentIntentsProphecy->reveal();
+
+        $container->set(DonationRepository::class, $donationRepoProphecy->reveal());
+        $container->set(EntityManagerInterface::class, $entityManagerProphecy->reveal());
+        $container->set(StripeClient::class, $stripeClientProphecy->reveal());
+
+        $requestPayload = $donation->toApiModel();
+        $requestPayload['autoConfirmFromCashBalance'] = true;
+        $request = $this->createRequest(
+            'PUT',
+            '/v1/donations/12345678-1234-1234-1234-1234567890ab',
+            json_encode($requestPayload),
+        )
+            ->withHeader('x-tbg-auth', DonationToken::create('12345678-1234-1234-1234-1234567890ab'));
+        $route = $this->getRouteWithDonationId('put', '12345678-1234-1234-1234-1234567890ab');
+
+        // act
+        $response = $app->handle($request->withAttribute('route', $route));
+
+        // assert
+        $payload = (string) $response->getBody();
+
+        $this->assertJson($payload);
+        $this->assertEquals(400, $response->getStatusCode());
+
+        $expectedPayload = new ActionPayload(400, ['error' => [
+            'type' => 'BAD_REQUEST',
+            'description' => 'Processing incomplete. Please refresh and check your donation funds balance',
+        ]]);
+        $expectedSerialised = json_encode($expectedPayload, JSON_PRETTY_PRINT);
+
+        $this->assertEquals($expectedSerialised, $payload);
+    }
+
+    public function testAddDataAutoconfirmHitsNoPaymentMethodException(): void
+    {
+        // arrange
+        $app = $this->getAppInstance();
+        /** @var Container $container */
+        $container = $app->getContainer();
+
+        $donation = $this->getTestDonation();
+        $donation = $this->prepareDonationBasics($donation);
+        $donation->setPaymentMethodType('customer_balance');
+
+        $donationRepoProphecy = $this->prophesize(DonationRepository::class);
+        $donationInRepo = $this->getTestDonation();  // Get a new mock object so DB has old values.
+        // Make it explicit that the payment method type is (the unsupported
+        // for auto-confirms) "card".
+        $donationInRepo->setPaymentMethodType('customer_balance');
+
+        $donationRepoProphecy
+            ->findAndLockOneBy(['uuid' => '12345678-1234-1234-1234-1234567890ab'])
+            ->willReturn($donationInRepo)
+            ->shouldBeCalledOnce();
+        $donationRepoProphecy
+            ->deriveFees(Argument::type(Donation::class), null, null)
+            ->willReturn($donation) // Actual fee calculation is tested elsewhere.
+            ->shouldBeCalledOnce();
```

Comment:
not a new issue, but noticing that the return value of deriveFees is always identical to the first argument - so it might be worth making it return void which would make it a little bit simpler to use and to test with.

Even better but for more work, I think deriveFees doesn't do any I/O so it wouldn't it make sense to hide it inside the Donation model and make the donation derive its own fees whenever necessary?

Reply:
I think it's true that we could return `void` but I'm not sure it unlocks a great simplification for any tests.

`deriveFees()` does set 2 mapped properties on Donation which we need denormalised after each change in card country/brand etc., so I'm not sure it would work to make it synthetic like that. Most fundamentally, we need to support changes in fee pricing over time.

Reply:
Not any great simplification, just a small simplification in many places - test and production code. I might make a separate PR to change it to void.

I see what you mean about about fee pricing over time - the `deriveFees` is run at a specific time, applying the policy we have at that time.